framework/aster-frame/src/sync/wait.rs

```diff
@@ -111,28 +113,54 @@ impl WaitQueue {
 
     /// Wake up one waiting thread.
     pub fn wake_one(&self) {
-        while let Some(waker) = self.wakers.lock_irq_disabled().pop_front() {
+        // Fast path
+        if self.is_empty() {
+            return;
+        }
+
+        loop {
+            let mut wakers = self.wakers.lock_irq_disabled();
+            let Some(waker) = wakers.pop_front() else {
+                break;
+            };
+            self.count.fetch_sub(1, Ordering::Relaxed);
+
             // Avoid holding lock when calling `wake_up`
+            drop(wakers);
             if waker.wake_up() {
-                return;
+                break;
             }
         }
     }
 
     /// Wake up all waiting threads.
     pub fn wake_all(&self) {
-        while let Some(waker) = self.wakers.lock_irq_disabled().pop_front() {
+        // Fast path
+        if self.is_empty() {
+            return;
+        }
+
+        loop {
+            let mut wakers = self.wakers.lock_irq_disabled();
+            let Some(waker) = wakers.pop_front() else {
+                break;
+            };
+            self.count.fetch_sub(1, Ordering::Relaxed);
+
             // Avoid holding lock when calling `wake_up`
+            drop(wakers);
             waker.wake_up();
         }
     }
 
     pub fn is_empty(&self) -> bool {
```

Comment:
Actually, this is more of an internal API that should not be exposed, especially after this PR. But maybe we want to deal with it in a separate PR?

Reply:
This PR is just a performance optimization,  so I choose to not making it an internal API.